solana/indexer/subscription.go

```diff
@@ -0,0 +1,219 @@
+package indexer
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gagliardetto/solana-go"
+	pb "github.com/rpcpool/yellowstone-grpc/examples/golang/proto"
+	"go.uber.org/zap"
+)
+
+// LaserStream from Helius only keeps the last 3000 slots.
+// Subtract 10 slots to be sure that the subscription doesn't fail.
+var MAX_SLOT_GAP = uint64(2990)
+
+type artistCoinsChangedNotification struct {
+	Operation string `json:"operation"`
+	NewMint   string `json:"new_mint"`
+	OldMint   string `json:"old_mint"`
+}
+
+func (s *SolanaIndexer) Subscribe(ctx context.Context) error {
+	conn, err := s.pool.Acquire(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to acquire database connection: %w", err)
+	}
+	defer conn.Release()
+
+	rawConn := conn.Conn()
+	_, err = rawConn.Exec(ctx, `LISTEN artist_coins_changed`)
+	if err != nil {
+		return fmt.Errorf("failed to listen for artist coins changes: %w", err)
+	}
+
+	defer func() {
+		s.logger.Info("received shutdown signal, stopping subscription")
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		coins, err := getArtistCoins(ctx, s.pool)
+		if err != nil {
+			return fmt.Errorf("failed to get artist coins: %w", err)
+		}
+		s.mintsFilter = append(coins,
+			s.config.SolanaConfig.MintAudio.String(),
+			s.config.SolanaConfig.MintUSDC.String(),
+		)
+
+		subscription, err := buildSubscriptionRequest(coins)
+		if err != nil {
+			return fmt.Errorf("failed to create subscription: %w", err)
+		}
+
+		lastIndexedSlot, err := withRetries(func() (uint64, error) {
+			return getCheckpointSlot(ctx, s.pool, subscription)
+		}, 5, time.Second*5)
+		if err != nil {
+			return fmt.Errorf("failed to get last indexed slot: %w", err)
+		}
+
+		latestSlot, err := withRetries(func() (uint64, error) {
+			return s.rpcClient.GetSlot(ctx, "confirmed")
+		}, 5, time.Second*5)
+		if err != nil {
+			return fmt.Errorf("failed to get slot: %w", err)
+		}
+
+		var fromSlot uint64
+		minimumSlot := latestSlot - MAX_SLOT_GAP
+		if lastIndexedSlot > minimumSlot {
+			fromSlot = lastIndexedSlot
+		} else {
+			if lastIndexedSlot == 0 {
+				s.logger.Warn("no last indexed slot found, starting from minimum slot and skipping backfill", zap.Uint64("fromSlot", minimumSlot))
+			} else {
+				s.logger.Warn("last indexed slot is too old, starting from minimum slot and backfilling", zap.Uint64("fromSlot", minimumSlot), zap.Uint64("toSlot", lastIndexedSlot))
+				go func(lastIndexedSlot, minimumSlot uint64) {
+					err := s.Backfill(ctx, lastIndexedSlot, minimumSlot)
```

Comment:
Maybe I missed something. It seems like we don't backfill artist coin data, just add a subscription and then re-subscribe whenever a new artist coin is added?